Tests/SessionTests.swift

```diff
@@ -972,6 +980,62 @@ class SessionTestCase: BaseTestCase {
             XCTFail("error should not be nil")
         }
     }
+
+    // MARK: Tests - Retry Policies
+
+    func testThatSessionRetriesRequestWhenRetryPolicyIsSet() {
+        // Given
+        let urlString = "https://httpbin.org/basic-auth/user/password"
+
+        let retryPolicies: [RetryPolicy] = [
+            MaxRetryCountPolicy(maxRetryCount: 0),
+            MaxRetryCountPolicy(maxRetryCount: 1),
+            MaxRetryCountPolicy(maxRetryCount: 2)
+        ]
+
+        let session = Session()
+
+        let expectation = self.expectation(description: "request should eventually fail")
+        var response: DataResponse<Any>?
+
+        // When
+        let request = session.request(urlString, retryPolicies: retryPolicies).validate().responseJSON { jsonResponse in
+            response = jsonResponse
+            expectation.fulfill()
+        }
+
+        waitForExpectations(timeout: timeout, handler: nil)
+
+        // Then
+        XCTAssertEqual(request.retryCount, 2)
+        XCTAssertEqual(response?.result.isSuccess, false)
+    }
+
+    func testThatSessionRetriesRequestWithBothRetryPoliciesAndRetrier() {
+        // Given
+        let urlString = "https://httpbin.org/basic-auth/user/password"
+        let handler = RequestHandler()
+        let retryPolicies: [RetryPolicy] = [MaxRetryCountPolicy(maxRetryCount: 1)]
+
+        let session = Session(adapter: handler, retrier: handler)
+
+        let expectation = self.expectation(description: "request should eventually fail")
+        var response: DataResponse<Any>?
+
+        // When
+        let request = session.request(urlString, retryPolicies: retryPolicies).validate().responseJSON { jsonResponse in
+            response = jsonResponse
+            expectation.fulfill()
+        }
```

Comment:
Pretty slick. Once the retry policies run out of retries, the interceptor kicks in and also allows one more retry.